## **Наука про дані в роздрібній торгівлі. Підготовка даних та базовий статистичний аналіз**

Орієнтовний час, необхідний для виконання: **45** хвилин 

## Анотація

Ця лабораторна робота присвячена завантаженню, попередній підготовці та базовому статистичному аналізу у сфері роздрібної торгівлі на основі даних про світові ціни на продукти харчування, отримані з сайту Всесвітньої продовольчої програми, що охоплюють такі продукти, як кукурудза, рис, боби, риба та цукор, для 76 країн та близько 1500 ринків.

## Вступ

Основна проблема, яку потрібно вирішити в цій лабораторній роботі - це завантаження, попередня підготовка та статистичний аналіз, включаючи візуалізацію набору даних.

Основна складність статистичного аналізу реальних даних полягає у тому, що вони готуються чи представляються у формі, не зручній для методів машинного навчання. Дуже часто реальні дані представляють собою є змішану інформацію, представлену в різних масштабах. Тому ці дані мають бути знайдені та перемасштабовані чи перераховані. У цій лабораторній роботі показані методи автоматичної попередньої підготовки реальних даних. Наступна проблема — уміння грамотно маніпулювати та перетворювати великі дані, щоб отримати зручний статистичний звіт як у табличному вигляді, так і у вигляді графіків.

## Матеріали та методи

У цій лабораторній роботі ми навчимося завантажувати заздалегідь підготовлені дані, виконувати базовий статистичний аналіз та візуалізувати дані.
Ця лабораторна робота складається з наступних кроків:
* Завантаження даних - завантаження та відображення даних із файлу
* Попередня підготовка даних - попередній аналіз структури даних, зміна структури даних та таблиць
* Статистичний аналіз - базовий статистичний аналіз
* Візуалізація даних - виведення різних даних на графіках
* Побудова лінії тренду - базовий метод побудови ліній тренду та прогнозу на його основі

Статистичні дані були отримані з https://www.kaggle.com/lasaljaywardena/global-food-prices-dataset. Цей набір даних має ліцензію CC BY-IGO, що дозволяє безкоштовне використання як у некомерційних, так і в комерційних цілях.

## Попередні вимоги
* [Python](https://www.python.org) — базовий рівень
* [Pandas](https://pandas.pydata.org)— базовий рівень 
* Statistics — базовий рівень

## Цілі

Після виконання цієї лабораторної роботи ви зможете:

* Завантажувати набір даних із файлів *.csv
* Проводити базовий аналіз даних
* Створювати нові та перераховувати значення існуючих стовпців
* Перетворити таблицю
* Візуалізувати дані за допомогою pandas та matplotlib
* Оцінювати
  * Мінімальне та максимальне значення
  * Середнє
  * Квартілі
* Будувати зведені таблиці

## Імпортування бібліотек/Означення допоміжних функцій

**Запуск поза Skills Network Labs** . Цей файл був протестований у Skills Network Labs. Запуск в іншому середовищі також можливий, але не гарантується і може вимагати іншої процедури налаштування.

Повинні бути встановлені бібліотеки Pandas, MatplotLib та SeaBorn.

In [1]:
# conda install -c conda-forge pandas

In [ ]:
# conda install -c conda-forge matplotlib

In [3]:
# conda install -c conda-forge seaborn 

## Завантаження даних із файлу *.csv

Спочатку треба імпортувати необхідні бібліотеки.

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

Давайте спочатку відключимо попередження за допомогою **[warnings.filterwarnings()](https://docs.python.org/3/library/warnings.html)**

In [2]:
import warnings
warnings.filterwarnings('ignore')

Далі задамо значення параметра точності при виведенні на екран, що дорівнює 2, щоб відображалися два десяткові знаки (замість 6 за замовчуванням) за допомогою **[pandas.set_option()](https://pandas.pydata.org/docs/reference/api/pandas.set_option.html)** та **[pd.options.display](https://pandas.pydata.org/pandas-docs/stable/user_guide/options.html)**.

In [4]:
pd.set_option("display.precision", 2)
pd.options.display.float_format = '{:.2f}'.format

Наступним кроком буде завантаження файлу даних з репозиторію за допомогою **[read_csv()](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html)**

In [5]:
df = pd.read_csv('global_food_prices.csv')

Тепер давайте вивчимо наш набір даних.

In [6]:
df

adm0_id    adm0_name  adm1_id   adm1_name  mkt_id    mkt_name  cm_id  \
0           1.00  Afghanistan      272  Badakhshan     266    Fayzabad     55   
1           1.00  Afghanistan      272  Badakhshan     266    Fayzabad     55   
2           1.00  Afghanistan      272  Badakhshan     266    Fayzabad     55   
3           1.00  Afghanistan      272  Badakhshan     266    Fayzabad     55   
4           1.00  Afghanistan      272  Badakhshan     266    Fayzabad     55   
...          ...          ...      ...         ...     ...         ...    ...   
2050633   271.00     Zimbabwe     3444    Midlands    5594  Mbilashaba    432   
2050634   271.00     Zimbabwe     3444    Midlands    5594  Mbilashaba    539   
2050635   271.00     Zimbabwe     3444    Midlands    5594  Mbilashaba    540   
2050636   271.00     Zimbabwe     3444    Midlands    5594  Mbilashaba    541   
2050637   271.00     Zimbabwe     3444    Midlands    5594  Mbilashaba    887   

                         cm_name  cur_id cur_name  pt_id pt_name  um_id  \
0                 Bread - Retail    0.00      AFN     15  Retail      5   
1                 Bread - Retail    0.00      AFN     15  Retail      5   
2                 Bread - Retail    0.00      AFN     15  Retail      5   
3                 Bread - Retail    0.00      AFN     15  Retail      5   
4                 Bread - Retail    0.00      AFN     15  Retail      5   
...                          ...     ...      ...    ...     ...    ...   
2050633   Beans (sugar) - Retail    0.00      ZWL     15  Retail      5   
2050634      Toothpaste - Retail    0.00      ZWL     15  Retail    116   
2050635    Laundry soap - Retail    0.00      ZWL     15  Retail      5   
2050636   Handwash soap - Retail    0.00      ZWL     15  Retail     66   
2050637  Fish (kapenta) - Retail    0.00      ZWL     15  Retail      5   

        um_name  mp_month  mp_year  mp_price  mp_commoditysource  
0            KG         1     2014     50.00                 NaN  
1            KG         2     2014     50.00                 NaN  
2            KG         3     2014     50.00                 NaN  
3            KG         4     2014     50.00                 NaN  
4            KG         5     2014     50.00                 NaN  
...         ...       ...      ...       ...                 ...  
2050633      KG         6     2021    233.33                 NaN  
2050634  100 ML         6     2021    112.50                 NaN  
2050635      KG         6     2021    114.00                 NaN  
2050636   250 G         6     2021     59.50                 NaN  
2050637      KG         6     2021   1200.00                 NaN  

[2050638 rows x 18 columns]

## Попередня підготовка даних

Вивчимо набір даних. Як бачите, він складається з 2050638 рядків × 18 стовпців. Також варто зазначити, що набір даних складається з різних типів інформації. Ми повинні бути впевнені, що Python правильно розпізнає ці типи. Для цього ми маємо використовувати **[pandas.info()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html?highlight=info#pandas.DataFrame.info)**.

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 511651 entries, 0 to 2050637
Data columns (total 11 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   adm0_name    511651 non-null  object 
 1   mkt_name     511651 non-null  object 
 2   cm_name      511651 non-null  object 
 3   cur_name     511651 non-null  object 
 4   pt_name      511651 non-null  object 
 5   um_name      511651 non-null  object 
 6   mp_month     511651 non-null  int64  
 7   mp_year      511651 non-null  int64  
 8   mp_price     511651 non-null  float64
 9   um_koef      510675 non-null  float64
 10  mp_price_un  510675 non-null  float64
dtypes: float64(3), int64(2), object(6)
memory usage: 46.8+ MB


In [32]:
df.describe()

mp_month   mp_year    mp_price   um_koef  mp_price_un
count 511651.00 511651.00   511651.00 510675.00    510675.00
mean       6.37   2016.00    12980.66      8.56      8455.21
std        3.40      4.46   205684.84     66.70    187673.01
min        1.00   1990.00        0.00      0.02         0.00
25%        3.00   2013.00       45.00      1.00        39.48
50%        6.00   2017.00      375.00      1.00       350.00
75%        9.00   2020.00     2700.00      1.00      2300.00
max       12.00   2021.00 21777780.00   1000.00  17250000.00

Дані включають країну, область, ринок, придбані товари, ціну та використану валюту, куплену кількість та місяць/рік покупки.

1. adm0_id: id країни
2. adm0_name: назва країни
3. adm1_id: id області
4. adm1_name: назва області
5. mkt_id: id ринку
6. mkt_name: назва ринку
7. cm_id: id закупленого товару
8. cm_name: закуплений товар
9. cur_id: id валюти
10. cur_name: назва валюти
11. pt_id: id типу ринку
12. pt_name: тип ринку (Роздрібний / Оптова торгівля / Виробник / Ферма)
13. um_id: id одиниць вимірювання
14. um_name: одиниця виміру товару
15. mp_month: місяць покупки
16. mpyear: рік покупки 
17. mpprice: заплачена вартість
18. mp_commoditysource: Джерело, що надає інформацію про ціни

Як бачите, набір даних складається з множини полів id. Це артефакт вилучення з бази даних. Ці поля не підходять для аналізу та мають бути виключені. Виберемо поля для аналізу

In [8]:
col = ['adm0_name', 'mkt_name', 'cm_name', 'cur_name', 'pt_name', 'um_name', 'mp_month', 'mp_year', 'mp_price']

І перетворимо набір даних:

In [9]:
df = df[col]
df

adm0_name    mkt_name                  cm_name cur_name pt_name  \
0        Afghanistan    Fayzabad           Bread - Retail      AFN  Retail   
1        Afghanistan    Fayzabad           Bread - Retail      AFN  Retail   
2        Afghanistan    Fayzabad           Bread - Retail      AFN  Retail   
3        Afghanistan    Fayzabad           Bread - Retail      AFN  Retail   
4        Afghanistan    Fayzabad           Bread - Retail      AFN  Retail   
...              ...         ...                      ...      ...     ...   
2050633     Zimbabwe  Mbilashaba   Beans (sugar) - Retail      ZWL  Retail   
2050634     Zimbabwe  Mbilashaba      Toothpaste - Retail      ZWL  Retail   
2050635     Zimbabwe  Mbilashaba    Laundry soap - Retail      ZWL  Retail   
2050636     Zimbabwe  Mbilashaba   Handwash soap - Retail      ZWL  Retail   
2050637     Zimbabwe  Mbilashaba  Fish (kapenta) - Retail      ZWL  Retail   

        um_name  mp_month  mp_year  mp_price  
0            KG         1     2014     50.00  
1            KG         2     2014     50.00  
2            KG         3     2014     50.00  
3            KG         4     2014     50.00  
4            KG         5     2014     50.00  
...         ...       ...      ...       ...  
2050633      KG         6     2021    233.33  
2050634  100 ML         6     2021    112.50  
2050635      KG         6     2021    114.00  
2050636   250 G         6     2021     59.50  
2050637      KG         6     2021   1200.00  

[2050638 rows x 9 columns]

In [10]:
df.mp_year.min()

1990

Давайте вивчимо інформацію стовпців набору даних.

Давайте створимо список країн у цьому наборі даних, використовуючи: **[pandas.Series.value_counts()](https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html)** та покажемо результати у вигляді таблиці **[pandas.Series.to_frame()](https://pandas.pydata.org/docs/reference/api/pandas.Series.to_frame.html)**

In [11]:
df['adm0_name'].value_counts().to_frame()

adm0_name
Rwanda                   137746
Bassas da India          137093
Syrian Arab Republic     116588
Philippines               82099
Mali                      73843
...                         ...
Viet Nam                    275
Azerbaijan                  125
Eritrea                     100
Georgia                      80
Venezuela                     6

[98 rows x 1 columns]

Як бачите, набір данних містить інформацію про 98 країн. Найбільше інформації про ринки Руанди, Індії та Сирійської Арабської Республіки.

Аналогічно визначаємо перелік ринків в цьому наборі даних.

In [12]:
df['mkt_name'].value_counts().to_frame()

mkt_name
National Average     19748
Bogota                5798
Khujand               5270
Dushanbe              5270
Gharm                 5241
...                    ...
Faya                     2
Socogim                  2
Vigwagwa                 1
Ali Addeh                1
Kalombo                  1

[3235 rows x 1 columns]

Як бачите, є інформація про 3235 ринків.

Давайте вивчимо типи куплених товарів: **[pandas.unique()](https://pandas.pydata.org/docs/reference/api/pandas.unique.html)**. Для спрощення аналізу ці списки мають бути впорядковані за допомогою **[list.sort()](https://docs.python.org/3/howto/sorting.html)**

In [13]:
u = list(df['cm_name'].unique())
u.sort()
u

['Alcohol (spray) - Retail',
 'Anchovies - Retail',
 'Apples (dried) - Retail',
 'Apples (red) - Retail',
 'Apples (red) - Wholesale',
 'Apples - Retail',
 'Apples - Wholesale',
 'Avocados - Retail',
 'Avocados - Wholesale',
 'Bananas (imported) - Retail',
 'Bananas (lakatan) - Retail',
 'Bananas (latundan) - Retail',
 'Bananas (local) - Retail',
 'Bananas (medium size) - Retail',
 'Bananas (saba) - Retail',
 'Bananas - Retail',
 'Bananas - Wholesale',
 'Barley (mixed) - Wholesale',
 'Barley (white) - Wholesale',
 'Barley - Retail',
 'Barley - Wholesale',
 'Batteries (big) - Retail',
 'Batteries (small) - Retail',
 'Beans (black) - Retail',
 'Beans (black) - Wholesale',
 'Beans (black, Nayarit) - Wholesale',
 'Beans (black, Veracruz) - Wholesale',
 'Beans (black, Zacatecas) - Wholesale',
 'Beans (black, imported) - Wholesale',
 'Beans (butter) - Retail',
 'Beans (catarino) - Retail',
 'Beans (cranberry) - Retail',
 'Beans (cranberry) - Wholesale',
 'Beans (dry) - Retail',
 'Beans (dry)

Як можна бачити, є багато видів покупок. Більше того, значення "Обмінний курс" використовується як проста покупка. Давайте вивчимо ці рядки, використовуючи зріз набору даних та функцію **[pandas.Series.str.contains()](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.contains.html )**

In [14]:
df[df['cm_name'].str.contains('Exchange rate')]

adm0_name   mkt_name                              cm_name cur_name  \
665      Afghanistan   Fayzabad               Exchange rate - Retail      AFN   
666      Afghanistan   Fayzabad               Exchange rate - Retail      AFN   
667      Afghanistan   Fayzabad               Exchange rate - Retail      AFN   
668      Afghanistan   Fayzabad               Exchange rate - Retail      AFN   
669      Afghanistan   Fayzabad               Exchange rate - Retail      AFN   
...              ...        ...                                  ...      ...   
1996823        Yemen  Al Jabeen  Exchange rate (unofficial) - Retail      YER   
1996824        Yemen  Al Jabeen  Exchange rate (unofficial) - Retail      YER   
1996825        Yemen  Al Jabeen  Exchange rate (unofficial) - Retail      YER   
1996826        Yemen  Al Jabeen  Exchange rate (unofficial) - Retail      YER   
1996827        Yemen  Al Jabeen  Exchange rate (unofficial) - Retail      YER   

        pt_name  um_name  mp_month  mp_year  mp_price  
665      Retail  USD/LCU         2     2014     57.55  
666      Retail  USD/LCU         7     2014     56.60  
667      Retail  USD/LCU         8     2014     56.62  
668      Retail  USD/LCU         9     2014     57.00  
669      Retail  USD/LCU        10     2014     57.38  
...         ...      ...       ...      ...       ...  
1996823  Retail  USD/LCU         1     2021    589.02  
1996824  Retail  USD/LCU         2     2021    592.54  
1996825  Retail  USD/LCU         3     2021    598.16  
1996826  Retail  USD/LCU         4     2021    593.24  
1996827  Retail  USD/LCU         6     2021    594.07  

[9970 rows x 9 columns]

Як ви можете бачити, ці рядки покупки складаєються з обмінного курсу для конкретного місяця та року для країн в одиницях місцевої валюти по відношенню до долара США. Ця ставка вказана у mp_price. Стовпець um_name містить скорочену назву валюти.

Також можна зазначити, що вартість покупок зазначається в одиницях місцевої валюти.

Це означає, що для порівняння ринкових цін різних країн ми маємо перерахувати стовпець mp_price у долари США. Це буде продемонстровано пізніше.

Давайте вивчимо, яка валюта використовується у різних країнах.

Для цього ми можемо використовувати **[pandas.crosstab()](https://pandas.pydata.org/docs/reference/api/pandas.crosstab.html)**

In [15]:
pd.crosstab(df["adm0_name"],
            df["cur_name"],
            )

cur_name       AFN    AMD   AOA  ARS  AZN  BDT  BIF  BOB  BTN  BYR  ...  UGX  \
adm0_name                                                           ...        
Afghanistan  15427      0     0    0    0    0    0    0    0    0  ...    0   
Algeria          0      0     0    0    0    0    0    0    0    0  ...    0   
Angola           0      0  1272    0    0    0    0    0    0    0  ...    0   
Argentina        0      0     0  990    0    0    0    0    0    0  ...    0   
Armenia          0  20600     0    0    0    0    0    0    0    0  ...    0   
...            ...    ...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
Venezuela        0      0     0    0    0    0    0    0    0    0  ...    0   
Viet Nam         0      0     0    0    0    0    0    0    0    0  ...    0   
Yemen            0      0     0    0    0    0    0    0    0    0  ...    0   
Zambia           0      0     0    0    0    0    0    0    0    0  ...    0   
Zimbabwe         0      0     0    0    0    0    0    0    0    0  ...    0   

cur_name      USD  VEF  VND  XAF  XOF    YER  ZAR    ZMW   ZWL  
adm0_name                                                       
Afghanistan     0    0    0    0    0      0    0      0     0  
Algeria         0    0    0    0    0      0    0      0     0  
Angola          0    0    0    0    0      0    0      0     0  
Argentina       0    0    0    0    0      0    0      0     0  
Armenia         0    0    0    0    0      0    0      0     0  
...           ...  ...  ...  ...  ...    ...  ...    ...   ...  
Venezuela       0    6    0    0    0      0    0      0     0  
Viet Nam        0    0  275    0    0      0    0      0     0  
Yemen           0    0    0    0    0  36806    0      0     0  
Zambia          0    0    0    0    0      0    0  42793     0  
Zimbabwe     4402    0    0    0    0      0    0      0  6555  

[98 rows x 84 columns]

Як бачите, це представлення даних не є зручним для аналізу. Краще використовувати **[pandas.DataFrame.groupby()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html?highlight=groupby#pandas.DataFrame.groupby)**

In [17]:
(df[df['um_name']=='USD/LCU']).groupby('adm0_name')['cur_name'].first().to_frame()

cur_name
adm0_name                                
Afghanistan                           AFN
Armenia                               AMD
Chad                                  XAF
Democratic Republic of the Congo      CDF
Ethiopia                              ETB
Iraq                                  IQD
Lebanon                               LBP
Liberia                               LRD
Libya                                 LYD
Somalia                               SOS
South Sudan                           SSP
Sudan                                 SDG
Syrian Arab Republic                  SYP
Tajikistan                            TJS
Venezuela                             VEF
Yemen                                 YER

Повернемося до вивчення типів закупівлі. Як ви могли бачити вище, існує безліч різновидів подібних видів покупок. Зокрема:

In [18]:
df[df['cm_name'].str.contains('Bread')].groupby('cm_name')['cm_name'].first().values

array(['Bread (bakery) - Retail',
       'Bread (bakery, parallel market) - Retail',
       'Bread (brotchen) - Retail', 'Bread (brown) - Retail',
       'Bread (common) - Retail', 'Bread (first grade flour) - Retail',
       'Bread (high grade flour) - Retail', 'Bread (khoboz) - Retail',
       'Bread (pita) - Retail', 'Bread (rye) - Retail',
       'Bread (shop) - Retail', 'Bread (traditional) - Retail',
       'Bread (vetkoek) - Retail', 'Bread (wheat) - Retail',
       'Bread - Retail'], dtype=object)

Залишимо в наборі даних інформацію про товари базового споживчого кошика та курс валют

In [19]:
df = df[df['cm_name'].str.contains('Bread') | 
        df['cm_name'].str.contains('Milk') | 
        df['cm_name'].str.contains('Meat') | 
        df['cm_name'].str.contains('Fish') | 
        df['cm_name'].str.contains('Rice') |
        df['cm_name'].str.contains('Water') |
        df['cm_name'].str.contains('Exchange rate')]
df

adm0_name    mkt_name                  cm_name cur_name pt_name  \
0        Afghanistan    Fayzabad           Bread - Retail      AFN  Retail   
1        Afghanistan    Fayzabad           Bread - Retail      AFN  Retail   
2        Afghanistan    Fayzabad           Bread - Retail      AFN  Retail   
3        Afghanistan    Fayzabad           Bread - Retail      AFN  Retail   
4        Afghanistan    Fayzabad           Bread - Retail      AFN  Retail   
...              ...         ...                      ...      ...     ...   
2050625     Zimbabwe      Matibi  Fish (kapenta) - Retail      ZWL  Retail   
2050626     Zimbabwe      Matibi  Fish (kapenta) - Retail      ZWL  Retail   
2050627     Zimbabwe      Matibi  Fish (kapenta) - Retail      ZWL  Retail   
2050628     Zimbabwe  Mbilashaba            Rice - Retail      ZWL  Retail   
2050637     Zimbabwe  Mbilashaba  Fish (kapenta) - Retail      ZWL  Retail   

        um_name  mp_month  mp_year  mp_price  
0            KG         1     2014     50.00  
1            KG         2     2014     50.00  
2            KG         3     2014     50.00  
3            KG         4     2014     50.00  
4            KG         5     2014     50.00  
...         ...       ...      ...       ...  
2050625      KG        12     2020   1187.50  
2050626      KG         2     2021   1123.33  
2050627      KG         3     2021   1390.00  
2050628      KG         6     2021    110.62  
2050637      KG         6     2021   1200.00  

[511651 rows x 9 columns]

Як бачите, кількість рядків зменшилася. Давайте дослідимо список видів покупок.

Замініть `##YOUR CODE GOES HERE##` на свій код Python.

In [ ]:
u = ##YOUR CODE GOES HERE##
u.sort()
u

<details><summary>Натисніть <b>тут,</b> щоб отримати розв'язок</summary> 
<code>    
u = list(df['cm_name'].unique())
u.sort()
u
    </code>
</details>

Згрупуємо покупки різних підтипів до однієї групи для подальшого аналізу. Для цього спочатку використайте **[lambda-function](https://www.geeksforgeeks.org/applying-lambda-functions-to-pandas-dataframe/)** та зрізи для видалення з назви покупок видів ринків. Вони відокремлюються від назви покупок за допомогою символів "-". Далі відкидаємо підтипи покупок, які починаються із символів "("

In [20]:
df.loc[:, 'cm_name'] = df['cm_name'].apply(lambda x : str(x)[:str(x).find(' -')] )
df.loc[:, 'cm_name'] = df['cm_name'].apply(lambda x : str(x) if str(x).find(' (') == -1 else str(x)[:str(x).find(' (')] )
df

adm0_name    mkt_name cm_name cur_name pt_name um_name  mp_month  \
0        Afghanistan    Fayzabad   Bread      AFN  Retail      KG         1   
1        Afghanistan    Fayzabad   Bread      AFN  Retail      KG         2   
2        Afghanistan    Fayzabad   Bread      AFN  Retail      KG         3   
3        Afghanistan    Fayzabad   Bread      AFN  Retail      KG         4   
4        Afghanistan    Fayzabad   Bread      AFN  Retail      KG         5   
...              ...         ...     ...      ...     ...     ...       ...   
2050625     Zimbabwe      Matibi    Fish      ZWL  Retail      KG        12   
2050626     Zimbabwe      Matibi    Fish      ZWL  Retail      KG         2   
2050627     Zimbabwe      Matibi    Fish      ZWL  Retail      KG         3   
2050628     Zimbabwe  Mbilashaba    Rice      ZWL  Retail      KG         6   
2050637     Zimbabwe  Mbilashaba    Fish      ZWL  Retail      KG         6   

         mp_year  mp_price  
0           2014     50.00  
1           2014     50.00  
2           2014     50.00  
3           2014     50.00  
4           2014     50.00  
...          ...       ...  
2050625     2020   1187.50  
2050626     2021   1123.33  
2050627     2021   1390.00  
2050628     2021    110.62  
2050637     2021   1200.00  

[511651 rows x 9 columns]

In [21]:
df['cm_name'].unique()

array(['Bread', 'Rice', 'Exchange rate', 'Milk', 'Meat', 'Fish',
       'Water spinach', 'Water', 'Watermelons'], dtype=object)

Як бачите, у нас лише дев'ять видів покупок, включаючи обмінний курс. Також у нас є додаткові покупки: "Water Spinach", "Watermelons". Вони залишилися, бо складаються зі слова «Water». Ми видалимо їх пізніше.

Дослідимо поле "Market name".

In [22]:
df['pt_name'].unique()

array(['Retail', 'Wholesale', 'Producer', 'Farm Gate'], dtype=object)

Як бачите, існує лише чотири типи ринків.

Перед аналізом ми повинні досліджувати одиниці виміру товару:

In [24]:
pd.crosstab(df["um_name"],
            df["cm_name"],
            )

cm_name      Bread  Exchange rate   Fish    Meat   Milk    Rice  Water  \
um_name                                                                  
0.8 KG           0              0      0       0    552       0      0   
1.1 KG        7913              0      0       0      0       0      0   
1.2 KG           0              0    526       0      0       0      0   
1.4 KG           0              0      0       0      0    1946      0   
1.5 KG           0              0    472       0      0       0      0   
1.5 L            0              0      0       0      0       0    825   
1.8 KG           0              0      0       0    711       0      0   
10 KG            0              0      0     364      0     108      0   
100 KG           0              0      0       0      0    8604      0   
100 Pounds       0              0      0       0      0      98      0   
120 KG           0              0      0       0      0      17      0   
125 G            0              0   1837       0      0       0      0   
150 G         1601              0      0       0      0       0      0   
160 G            0              0   2818       0      0       0      0   
170 G            0              0    636       0      0       0      0   
185 G            0              0   1901       0      0       0      0   
2 KG             0              0      0       0      0       6      0   
2.1 KG           0              0      0       0      0       0      0   
20 G             0              0      0       0   1007       0      0   
20 KG            0              0    161       0      0       0      0   
200 G            0              0   1404    2136      0       0      0   
200 ML           0              0      0       0   1367       0      0   
25 KG            0              0      0       0      0     917      0   
380 G            0              0      0       0     68       0      0   
385 G            0              0      0       0    158       0      0   
400 G          178              0      0       0     94       0      0   
45 KG            0              0      0       0      0    1789      0   
46 KG            0              0      0     687      0    3619      0   
5 KG             0              0      0       0      0     226      0   
5 pcs         1407              0      0       0      0       0      0   
50 KG            0              0      0       0      0    8683      0   
50 Pounds        0              0      0       0    156       0      0   
500 G          183              0      0       0    590     504      0   
500 ML           0              0      0       0    412       0      0   
700 G          648              0      0       0      0       0      0   
800 G           16              0      0       0      0       0      0   
84 KG            0              0      0       0      0     355      0   
90 KG            0              0      0       0      0     708      0   
900 G            0              0      0       0   2124     156      0   
Bundle           0              0    381       0      0       0      0   
Cuartilla        0              0      0       0      0     234      0   
Cubic meter      0              0      0       0      0       0   1070   
KG           10868              0  44071  113250   3797  216027      0   
L                0              0      0       0  24062       0   1138   
Libra            0              0    253     185      0     727      0   
Loaf          3675              0      0       0      0       0      0   
MT               0              0      0       0      0    2178      0   
Marmite          0              0      0       0      0    1730      0   
Pound          418              0    267     861     84    2037      0   
USD/LCU          0           9970      0       0      0       0      0   
Unit          8835              0     24     977    808       0      0   

cm_name      Water spinach  Watermelons  
um_name                            

Як бачите, у наборі даних є багато різних одиниць виміру. Для статистичної обробки потрібно розрахувати вартість конкретної покупки в еталонних одиницях. Для такого еталону виберемо 1 кг.
Для цього створимо словник ключів, що складається з ключа (тип одиниці) – вартість (коефіцієнт пропорційності до 1 кг). Курс обміну звичайно міняти не будемо - для цього використовуватимемо значення, що дорівнює 1 (останній рядок у словнику).

In [25]:
d = {'KG': 1,
 'Unit': 1,
 'Pound': 0.45,
 '400 G': 0.4,
 '150 G': 0.15,
 '5 pcs': 1,
 '500 G': 0.5,
 '700 G': 0.7,
 '800 G': 0.8,
 '1.1 KG': 1.1,
 'Libra': 0.45,
 '900 G': 0.9,
 '100 KG': 100,
 '46 KG': 46,
 'Cuartilla': 2.88,
 '90 KG': 90,
 '120 KG': 120,
 '50 KG': 50,
 'MT': 1000,
 '45 KG': 45,
 '84 KG': 84,
 '100 Pounds': 45,
 'Marmite': 1,
 '5 KG': 5,
 '10 KG': 10,
 '1.4 KG': 1.4,
 '25 KG': 25,
 '2 KG': 2,
 'Loaf' : 1,
 '50 Pounds': 22.5,
 '380 G': 0.38,
 '500 ML': 0.5,
 '385 G': 0.385,
 '0.8 KG': 0.8,
 '200 ML': 0.2,
 '20 G': 0.02,
 '1.8 KG': 1.8,
 '125 G': 0.125,
 '20 KG': 20,
 '170 G': 0.17,
 '185 G': 0.185,
 '200 G': 0.2,
 '1.5 KG': 1.5,
 '1.2 KG': 1.2,
 'Bundle': 46,
 '160 G': 0.16,
 'Cubic meter': 1,
 '1.5 L': 1.5,
 'L': 1,
 'USD/LCU': 1}

**[pandas.Series.map()](https://pandas.pydata.org/docs/reference/api/pandas.Series.map.html)** можна використовувати для **заміни значень у стовпці**, передавши його як словник аргументів у формі `{old_value: new_value}`.

Давайте застосуємо цей словник для розрахунку нового стовпця з коефіцієнтами. Потім використовуємо цю колонку для розрахунку ціни купівлі за 1 кг.

In [26]:
df['um_koef'] = df['um_name'].map(d)
df['mp_price_un'] = df['mp_price'] / df['um_koef']

In [27]:
df

adm0_name    mkt_name cm_name cur_name pt_name um_name  mp_month  \
0        Afghanistan    Fayzabad   Bread      AFN  Retail      KG         1   
1        Afghanistan    Fayzabad   Bread      AFN  Retail      KG         2   
2        Afghanistan    Fayzabad   Bread      AFN  Retail      KG         3   
3        Afghanistan    Fayzabad   Bread      AFN  Retail      KG         4   
4        Afghanistan    Fayzabad   Bread      AFN  Retail      KG         5   
...              ...         ...     ...      ...     ...     ...       ...   
2050625     Zimbabwe      Matibi    Fish      ZWL  Retail      KG        12   
2050626     Zimbabwe      Matibi    Fish      ZWL  Retail      KG         2   
2050627     Zimbabwe      Matibi    Fish      ZWL  Retail      KG         3   
2050628     Zimbabwe  Mbilashaba    Rice      ZWL  Retail      KG         6   
2050637     Zimbabwe  Mbilashaba    Fish      ZWL  Retail      KG         6   

         mp_year  mp_price  um_koef  mp_price_un  
0           2014     50.00     1.00        50.00  
1           2014     50.00     1.00        50.00  
2           2014     50.00     1.00        50.00  
3           2014     50.00     1.00        50.00  
4           2014     50.00     1.00        50.00  
...          ...       ...      ...          ...  
2050625     2020   1187.50     1.00      1187.50  
2050626     2021   1123.33     1.00      1123.33  
2050627     2021   1390.00     1.00      1390.00  
2050628     2021    110.62     1.00       110.62  
2050637     2021   1200.00     1.00      1200.00  

[511651 rows x 11 columns]

Ми підготували набір даних для статистичного аналізу.

У цій лабораторній роботі ми проаналізуємо значення та динаміку цін покупок споживчого кошика у різних країнах.
Для цього ми повинні згрупувати рядки за країнами, роком, місяцем та ринками. Після цього треба перетворити таблицю, створивши нові окремі стовпці для кожного типу покупок кошика клієнтів.

Ми можемо зробити це за допомогою **[pivot_table()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.pivot_table.html?highlight=pivot_table#pandas.DataFrame.pivot_table)**

In [28]:
df_res = pd.pivot_table(df, values= 'mp_price', index= ['adm0_name', 'mp_year', 'mp_month', 'mkt_name'], columns=['cm_name'], aggfunc='mean', margins=False)

In [29]:
df_res

cm_name                                               Bread  Exchange rate  \
adm0_name   mp_year mp_month mkt_name                                        
Afghanistan 2000    1        Hirat                    15.63            NaN   
                             Jalalabad                14.86            NaN   
                             Kabul                    14.26            NaN   
                             Kandahar                 16.25            NaN   
                    2        Hirat                    15.63            NaN   
...                                                     ...            ...   
Zimbabwe    2021    6        Sizinda Shopping Centre    NaN            NaN   
                             Torwood 1                  NaN            NaN   
                             Torwood 2                  NaN            NaN   
                             Victoria Falls             NaN            NaN   
                             Zvishavane Urban           NaN            NaN   

cm_name                                                 Fish  Meat  Milk  \
adm0_name   mp_year mp_month mkt_name                                      
Afghanistan 2000    1        Hirat                       NaN   NaN   NaN   
                             Jalalabad                   NaN   NaN   NaN   
                             Kabul                       NaN   NaN   NaN   
                             Kandahar                    NaN   NaN   NaN   
                    2        Hirat                       NaN   NaN   NaN   
...                                                      ...   ...   ...   
Zimbabwe    2021    6        Sizinda Shopping Centre     NaN   NaN   NaN   
                             Torwood 1               1000.00   NaN   NaN   
                             Torwood 2                   NaN   NaN   NaN   
                             Victoria Falls          1270.45   NaN   NaN   
                             Zvishavane Urban        1227.70   NaN   NaN   

cm_name                                                Rice  Water  \
adm0_name   mp_year mp_month mkt_name                                
Afghanistan 2000    1        Hirat                      NaN    NaN   
                             Jalalabad                  NaN    NaN   
                             Kabul                      NaN    NaN   
                             Kandahar                   NaN    NaN   
                    2        Hirat                      NaN    NaN   
...                                                     ...    ...   
Zimbabwe    2021    6        Sizinda Shopping Centre  82.50    NaN   
                             Torwood 1                85.00    NaN   
                             Torwood 2                85.00    NaN   
                             Victoria Falls          123.43    NaN   
                             Zvishavane Urban        152.44    NaN   

cm_name                                               Water spinach  \
adm0_name   mp_year mp_month mkt_name                                 
Afghanistan 2000    1        Hirat                              NaN   
                             Jalalabad                          NaN   
                             Kabul                              NaN   
                             Kandahar                           NaN   
                    2        Hirat                              NaN   
...                                                             ...   
Zimbabwe    2021    6        Sizinda Shopping Centre            NaN   
                             Torwood 1                          NaN   
                             Torwood 2                          NaN   
                             Victoria Falls                     NaN   
                             Zvishavane Urban                   NaN   

cm_name                                               Watermelons  
adm0_name   mp_year mp_month mkt_name                              
Afghanistan 2000    1    

Як бачите, новий набір даних складається з 170 510 рядків. 

Тепер ми можемо легко видалити стовпці: Water Spinach і Watermelons.

In [30]:
df_res.columns

Index(['Bread', 'Exchange rate', 'Fish', 'Meat', 'Milk', 'Rice', 'Water',
       'Water spinach', 'Watermelons'],
      dtype='object', name='cm_name')

In [ ]:
df_res = ##YOUR CODE GOES HERE##
df_res

<details><summary>Натисніть <b>тут,</b> щоб отримати розв'язок</summary> 
<code>    
df_res = df_res[['Exchange rate', 'Bread', 'Fish', 'Meat', 'Milk', 'Rice', 'Water']]
df_res
    </code>
</details>

## Статистичний аналіз

### Основи статистичного аналізу

Як ви можете бачити, результуючий набір даних має зручну для аналізу структуру. Сгруповані стовпці представлені як індекси. Однак для статистичного аналізу ці стовпці необхідно перетворити на вигляд звичайних стовпців. Ми можемо зробити це за допомогою **[pandas.DataFrame.reset_index()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.reset_index.html)**

In [ ]:
df_stat = df_res.reset_index()
df_stat

Давайте визначимо, скільки записів доступно для різних країн у цьому наборі даних у абсолютному та нормованому вигляді.

In [ ]:
df_stat["adm0_name"].value_counts().to_frame()

In [ ]:
df_stat["adm0_name"].value_counts(normalize=True).to_frame()

Як бачите, нам доступна інформація про 93 країни. Також найбільшу кількість записів становлять Малі, Індія та Нігер.

Створимо список наявних країн:

In [ ]:
##YOUR CODE GOES HERE##

<details><summary>Натисніть <b>тут,</b> щоб отримати розв'язок</summary> 
<code>    
df_stat["adm0_name"].unique()
    </code>
</details>

Давайте розрахуємо базову статистичну інформацію про покупки в Малі, використовуючи **[pandas.describe()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html?highlight=describe#pandas.DataFrame.describe)**.

In [ ]:
df_stat[df_stat['adm0_name']=='Mali'].describe()

У цій таблиці можна побачити консолідовану статистичну інформацію, таку як кількість непустих рядків, середнє значення, стандартне відхилення, максимум, мінімум і т.д.
Як бачите, хліб та вода не користуються популярністю в Індії.

Звичайно, ви можете розрахувати будь-який статистичний коефіцієнт, використовуючи **[statistics](https://docs.python.org/3/library/statistics.html)** або **[numpy](https://numpy.org )**

Давайте порахуємо те саме для України

In [ ]:
##YOUR CODE GOES HERE##

<details><summary>Натисніть <b>тут,</b> щоб отримати розв'язок</summary> 
<code>    
df_stat[df_stat['adm0_name']=='Ukraine'].describe()
    </code>
</details>

Як бачите, українці надають перевагу хлібу, натомість риба не є популярною.

Давайте дослідимо Китай

In [ ]:
##YOUR CODE GOES HERE##

<details><summary>Натисніть <b>тут,</b> щоб отримати розв'язок</summary> 
<code>    
df_stat[df_stat['adm0_name']=='China'].describe()
    </code>
</details>

Як бачите, Китай містить інформацію лише про рис.

А що на раххунок Індії?

In [ ]:
##YOUR CODE GOES HERE##

<details><summary>Натисніть <b>тут,</b> щоб отримати розв'язок</summary> 
<code>    
df_stat[df_stat['adm0_name']=='Bassas da India'].describe()
    </code>
</details>

Індія містить інформацію про рис і молоко.

Вивчимо динаміку закупівельних цін в Індії. Для цього ми повинні відфільтрувати дані та згрупувати їх за роками:

In [ ]:
d = df_stat[df_stat['adm0_name']=='Bassas da India'].groupby('mp_year').mean()
d

### Візуалізація

Тепер ми можемо відобразити цю інформацію за допомогою **[pandas.DataFrame.plot()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.plot.html?highlight=plot#pandas.DataFrame.plot)**

In [ ]:
d[['Milk','Rice']].plot()

Як бачите, ціни на рис в Індії нестабільні та мають коливання кожні десять років. Ціни на молоко зростають з кожним роком.

Давайте зробимо те саме для України:

In [ ]:
d = df_stat[df_stat['adm0_name']=='Ukraine'].groupby('mp_year').mean()
d

In [ ]:
d[['Bread', 'Meat', 'Milk', 'Rice']].plot()

Як бачите найдорожчі ціни на м'ясо. Коливань цін не спостерігається. Також видно, що з кожним роком ціни зростають. Причиною може бути інфляція.

Подивимося на динаміку цін на рис у Китаї:

In [ ]:
d = ##YOUR CODE GOES HERE##
d

<details><summary>Натисніть <b>тут,</b> щоб отримати розв'язок</summary> 
<code>    
d = df_stat[df_stat['adm0_name']=='China'].groupby('mp_year').mean()
d
    </code>
</details>

In [ ]:
##YOUR CODE GOES HERE##

<details><summary>Натисніть <b>тут,</b> щоб отримати розв'язок</summary> 
<code>    
d[['Rice']].plot()
    </code>
</details>

Як бачите, останніми роками ціни падають. Від 4,5 юанів до 4,10 юанів

Давайте досліджуємо середні ціни на різних китайських ринках:

In [ ]:
d = df_stat[df_stat['adm0_name']=='China'].groupby(['mkt_name']).mean()
print(d)
d['Rice'].plot.bar()

Як бачите, Yuncheng – найдорожчий ринок, Hubei – найдешевший.

Ви також можете побудувати його за допомогою кругової діаграми:

In [ ]:
d['Rice'].plot.pie()

Також ми можемо досліджувати, як змінювалася структура ринкових цін за роками:

In [ ]:
d = df_stat[df_stat['adm0_name']=='China']
d = d[['mkt_name', 'mp_year', 'Rice']]
d = pd.pivot_table(d, values= 'Rice', index= ['mp_year'], columns='mkt_name', aggfunc='mean', margins=False)
print(d)
d.plot.bar()

Порівняємо ринкові ціни різних країн.

Як згадувалося раніше, набір даних включає країни, де національною валютою є долари США. Подивимося на них:

In [ ]:
usd_countries = list((df[(df['cur_name']=='USD')])['adm0_name'].unique())
usd_countries

Для цих країн поле "Exchange rate" порожнє. Тому для порівняння цін нам потрібно змінити значення цього стовпця на одиницю:

In [ ]:
df_stat.loc[:, 'Exchange rate'] = df_stat.apply(lambda x : 1 if x['adm0_name'] in usd_countries else x['Exchange rate'], axis=1 )

Перед розрахунком цін у доларах США ми видаляємо всі рядки, що містять порожні значення у стовпці «Exchange rate», використовуючи **[pandas.DataFrame.dropna()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html)**:

In [ ]:
df_stat_multy = df_stat[df_stat['Exchange rate'].notna()]
df_stat_multy

Давайте отримаємо список країн, де був відомий обмінний курс:

In [ ]:
df_stat_multy["adm0_name"].value_counts().to_frame()

In [ ]:
len(df_stat_multy["adm0_name"].value_counts())

Як бачите, залишилося лише 23 країни. Перерахуємо ціни на рис у доларах США

In [ ]:
df_stat_multy['Rice_USD'] = df_stat_multy['Rice'] / df_stat_multy['Exchange rate']
df_stat_multy

Потім ми можемо порівняти динаміку цін на рис у різних країнах:

In [ ]:
df_stat_multy = pd.pivot_table(df_stat_multy, values= 'Rice_USD', index= ['mp_year'], columns='adm0_name', aggfunc='mean', margins=False)
df_stat_multy

In [ ]:
len(df_stat_multy.columns)

Як бачите, лише 17 країн із 23 містять інформацію про ціну на рис.

Порівняємо динаміку цін для Конго, Сомалі, Таджикистану та Еквадору.

In [ ]:
df_stat_multy = df_stat_multy[['Democratic Republic of the Congo','Somalia', 'Tajikistan', 'Ecuador']]

In [ ]:
df_stat_multy.plot()

### Лінія тренду

Давайте порахуємо тренд ціни на рис для вказаних країн. Для цього ми маємо:
1. Вилучити тимчасовий ряд з DataFrame для досліджуваної країни.
2. Відкинути значення NaN
3. Вилучити значення **x** (року) та **y** (ціна на рис) з часового ряду і перетворити їх на вектор.
4. Створити **[sklearn.linear_model.LinearRegression()](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html)** модель
5. Підігнати модель за значеннями **x** та **y**
6. Побудувати прогноз на наступні 3 роки
7. Відобразити результат на графіку

In [ ]:
for c in df_stat_multy.columns:
    y = df_stat_multy[c].dropna()
    x = y.index
    from sklearn.linear_model import LinearRegression
    model = LinearRegression()
    x = np.reshape(x, (-1, 1)) # transform X values
    y = np.reshape(y.values, (-1, 1)) # transform Y values
    model.fit(x, y)
    X_pred= np.append(x, [2022, 2023, 2024])
    X_pred = np.reshape(X_pred, (-1, 1))
    # calculate trend
    trend = model.predict(X_pred)
    plt.figure()
    plt.plot(X_pred, trend, "-", label='Trend')
    plt.plot(x, y, ".", label='Real data')
    plt.xlabel('Year')
    plt.ylabel('Price in USD')
    plt.legend(title=c)
    plt.show()

Як видно з рисунків, лінійна модель не є точною, але дає змогу розрахувати основний тренд цін. Для кращого прогнозування слід використовувати різні підходи. Це буде показано у наступних лабораторних роботах.

## Висновки

Як показує практика, дані, отримані у реальних натурних експериментах, не придатні для прямої статистичної обробки. Тому в цій лабораторній роботі ми вивчили основні методи завантаження та попередньої підготовки даних. Ми вивчили, як завантажувати, перетворювати, групувати, перераховувати та візуалізувати дані.
На відміну від добре відомих класичних підходів до статистичного аналізу даних, Python містить безліч потужних бібліотек, що дозволяють легко та швидко маніпулювати даними. Тому ми вивчили основні методи автоматизації з бібліотеки Pandas для попередньої підготовки та статистичного аналізу даних. Також ми вивчили основні методи візуалізації даних за допомогою бібліотеки MatplotLib, яка містить ефективні засоби візуального аналізу даних.

## Завдання

1. Оберіть дані, де виконується умова вашого варінту і 
виведіть загальну кількість даних згідно проміжку з вашого варіанту:
- 2000 < mp_year < 2015
- 2005 < mp_year < 2010
- 1994 < mp_year < 2007
- 1997 < mp_year < 2005
- 2010 < mp_year < 2020
- 2002 < mp_year < 2015
- 1994 < mp_year < 2011
- 2012 < mp_year < 2021
- 2001 < mp_year < 2021
- 1995 < mp_year < 2011

2. Перевірте чи поле mp_price даних є добре розпреділеними за допомогою гістограм.
Зробіть гістограми надто детальні і ті, які містять недостатньо інформації.
Підберіть найкращу кількість bins для гістограм.

3. Побудуйте scatter plots по полях mp_year і mp_price.

4. Згрупуйте дані згідно критеріїв вашого варіанту і збережіть результат в нову колонку даних. Побудуйте таблицю, як в прикладі нижче. Друге поле даних,
яке ви будете розглядати - pt_name. Можна застосувати crosstab.
- mp_year <= 2000, 2000 < mp_year < 2015, mp_year >= 2015
- mp_year <= 2005, 2005 < mp_year < 2010, mp_year >= 2010
- mp_year <= 1994, 1994 < mp_year < 2007, mp_year >= 2007
- mp_year <= 1997, 1997 < mp_year < 2005, mp_year >= 2005
- mp_year <= 2010, 2010 < mp_year < 2020, mp_year >= 2020
- mp_year <= 2002, 2002 < mp_year < 2015, mp_year >= 2015
- mp_year <= 1994, 1994 < mp_year < 2011, mp_year >= 2011
- mp_year <= 2012, 2012 < mp_year < 2021, mp_year >= 2021
- mp_year <= 2001, 2001 < mp_year < 2021, mp_year >= 2021
- mp_year <= 1995, 1995 < mp_year < 2011, mp_year >= 2011

Орієнтовний вид фінальної таблиці нижче

mp_year_sorted  above 2015  between 2000 and 2015  under 2000
pt_name                                                      
Retail                0.71                   0.28        0.00
Wholesale             0.51                   0.49        0.00
All                   0.69                   0.31        0.00

5. Знайдіть максимальне, мінімальне, середнє, популярне, дисперсію випадкової величини, верхню і нижню квартілі значення поля mp_year.

6. Вирішіть проблему пропущених значень одним із способом на ваш вибір:
- adm0_name
- adm1_name
- cm_name
- mkt_name
- pt_name
- adm0_name
- adm1_name
- cm_name
- mkt_name
- pt_name

7. Текстові змінні не можуть використовуватися в подальшому для побудови регресій чи інших 
математичних моделей. Є кілька способів подолати цю проблему. Перший спосіб - присвоїти відповідні
числові значення - від 1 до n. Другий підхід - використати даміс змінні. Застосуйте один з
цих підходів для свого варіанту:
- adm0_name
- adm1_name
- cm_name
- mkt_name
- pt_name
- adm0_name
- adm1_name
- cm_name
- mkt_name
- pt_name

8. Побудуйте scatter plot змінних - mp_year vs:
- adm0_name
- adm1_name
- cm_name
- mkt_name
- pt_name
- adm0_name
- adm1_name
- cm_name
- mkt_name
- pt_name


9. Створіть нову колонку у даних, значення якої обчислюється log від mp_year.

## Автори

Розробник: [Yaroslav Vyklyuk, prof., PhD., DrSc](http://vyklyuk.bukuniver.edu.ua/en/)

Консультант з роздрібної торгівлі: [Olha Vdovichena, ass. prof, PhD](https://scholar.google.ru/citations?user=3vIQ33YAAAAJ&hl=uk)

 Copyright &copy; 2020 IBM Corporation. This notebook and its source code are released under the terms of the [MIT License](https://cognitiveclass.ai/mit-license/).